test_app/Gemfile

```diff
@@ -0,0 +1,53 @@
+source 'https://rubygems.org'
+git_source(:github) { |repo| "https://github.com/#{repo}.git" }
+
+ruby '2.7.5'
+
+# Bundle edge Rails instead: gem 'rails', github: 'rails/rails', branch: 'main'
+gem 'rails', '~> 6.1.4', '>= 6.1.4.4'
```

Comment:
What are all the variations of `~>`, `>=` etc and what are their meaning?

Reply:
The example from https://bundler.io/gemfile.html shows three most common version specifier:

```
1. - gem 'rails', '5.0.0'
2. - gem 'rack',  '>=1.0'
3. - gem 'thin',  '~>1.1'
```

The first example specifies the rails gem to use the version of gem "rails", in the example we are only using the particular version 5.0.0.
The second example ">=1.0" indicate that any version later than 1.0 is acceptable. There is no upper bound.
The third example '~>1.1' can be interpreted in this way ">=1.1", "<1.2", this is because we worry the changes in 1.2 ( in this example will break the program, hence we only accept the range between 1.1 to 1.2 (not include 1.2)